python/cudf/cudf/core/udf/typing.py

```diff
@@ -0,0 +1,186 @@
+            return MaskedType(other.literal_type)
+
+
+@typeof_impl.register(_NAType)
+def typeof_na(val, c):
+    '''
+    Tie instances of _NAType (cudf.NA) to our NAType.
+    Effectively make it so numba sees `cudf.NA` as an
+    instance of this NAType -> handle it accordingly.
+    '''
+    return NAType()
+
+register_model(NAType)(models.OpaqueModel)
+
+
+# Ultimately, we want numba to produce PTX code that specifies how to add
+# two singular `Masked` structs together, which is defined as producing a
+# new `Masked` with the right validity and if valid, the correct value. 
+# This happens in two phases:
+#   1. Specify that `Masked` + `Masked` exists and what it should return
+#   2. Implement how to actually do (1) at the LLVM level
+# The following code accomplishes (1) - it is really just a way of specifying
+# that the `+` operation has a CUDA overload that accepts two `Masked` that
+# are parameterized with `value_type` and what flavor of `Masked` to return.
+@cuda_decl_registry.register_global(operator.add)
+class MaskedScalarAdd(AbstractTemplate):
+    def generic(self, args, kws):
+        '''
+        Typing for `Masked` + `Masked`
+        Numba expects a valid numba type to be returned if typing is successful
+        else `None` signifies the error state (this is common across numba)
+        '''
+        if isinstance(args[0], MaskedType) and isinstance(args[1], MaskedType):
+            # In the case of op(Masked, Masked), the return type is a Masked
+            # such that Masked.value is the primitive type that would have 
+            # been resolved if we were just adding the `value_type`s. 
+            return_type = self.context.resolve_function_type(
+                self.key, (args[0].value_type, args[1].value_type), kws
+            ).return_type
+            return nb_signature(
+                MaskedType(return_type),
+                MaskedType(args[0].value_type),
+                MaskedType(args[1].value_type),
+            )
+
+@cuda_decl_registry.register_global(operator.add)
+class MaskedScalarAddNull(AbstractTemplate):
+    def generic(self, args, kws):
+        '''
+        Typing for `Masked` + `NA`
+        Handles situations like `x + cudf.NA`
+        '''
+        if isinstance(args[0], MaskedType) and isinstance(args[1], NAType):
+            # In the case of op(Masked, NA), the result has the same
+            # dtype as the original regardless of what it is
+            return_type = args[0].value_type
+            return nb_signature(
+                MaskedType(return_type),
+                MaskedType(args[0].value_type),
+                NAType(),
+            )
+
+@cuda_decl_registry.register_global(operator.add)
+class MaskedScalarAddConstant(AbstractTemplate):
+    def generic(self, args, kws):
+        '''
+        Typing for `Masked` + a constant literal
+        handles situations like `x + 1`
+        '''
+        if isinstance(args[0], MaskedType) and isinstance(
```

Comment:
This might also need to handle the arguments the other way around, where `args[1]` is the masked type.

Reply:
This should be fixed